File: perception/noisy_image_eliminate.py
#!/usr/bin/env python

'''
WPI Team Capricorn
NASA Space Robotics Challenge Phase 2

Author: Mahimana Bhatt
Team Endurance

This code reads stereo data from the topic, removes noisy images for the video stream and published it on new topics:
/capricorn/left
/capricorn/right
 
Algorithm: This algorithm first extracts edges from the image with high threshold, this operation leaves only noise and 
some terrain edges in a binary form, we calculate the sum of the whole image, in a noisy image the white pixels will be 
way greater than a non-noisy image, hence, we calculate the difference between adjacent edge image sums, and if the 
difference is greater than a threshold, the image is considered as noisy and we don't update the last sum for the noisy 
image to avoid two adjacent noisy images to pass through the filter.

Improvements:
1. When the base station enters into the image, the number of white pixels in edge image increases which causes
a delay to consider as a normal image, this needs improvement.
2. The threshold of the image can be made adaptive to increase the elimination of noisy images.
'''

#TODO: Convert code in CPP for optimization
#TODO: do both of the left and right check in two separate functions

# Computer Vision Libraries
from cv2 import cv2
import numpy as np

import sys
import logging
# ROS Libraries
import rospy
import message_filters
from cv_bridge import CvBridge
from sensor_msgs.msg import Image, CameraInfo

class noisy_image_eliminate:
    bridge = CvBridge()
    right_last_sum = 0
    left_last_sum = 0
    right_actual_sum = 0
    left_actual_sum = 0
    threshold = 4000
    upper_threshold = 1.3
    lower_threshold = 0.7

    def check_left(self, left):
        left_image = self.bridge.imgmsg_to_cv2(left,desired_encoding='passthrough')
        left_edge = cv2.Canny(left_image, 250, 255)

        left_sum = int(np.sum(left_edge))

        left_diff = left_sum-self.left_last_sum

        if(left_sum >= self.left_actual_sum * self.lower_threshold and left_sum <= self.left_actual_sum * self.upper_threshold):
            # print("Corrected")
            self.left_last_sum = left_sum

        self.left_actual_sum = left_sum

        # For debugging
        # print("Left sum: ", left_sum,"diff: ", left_diff)

        if(left_diff < self.threshold):
            self.left_last_sum = left_sum
            return True
        else:
            return False

    def check_right(self, right):
        right_image = self.bridge.imgmsg_to_cv2(right,desired_encoding='passthrough')
        right_edge = cv2.Canny(right_image, 250, 255)

        right_sum = int(np.sum(right_edge))

        right_diff = right_sum-self.right_last_sum

        if(right_sum >= self.right_actual_sum * self.lower_threshold and right_sum <= self.right_actual_sum * self.upper_threshold):
            # print("Corrected")
            self.right_last_sum = right_sum

        self.right_actual_sum = right_sum

        # For debugging
        # print("Right sum: ", right_sum,"diff: ", right_diff)

        if(right_diff < self.threshold):
            self.right_last_sum = right_sum
            return True
        else:
            return False


    def loop(self,right, right_info, left, left_info):
        if(self.check_left(left)):
            self.left_image_pub.publish(left)  
            self.left_info_pub.publish(left_info)    

        if(self.check_right(right)):
            self.right_image_pub.publish(right)  
            self.right_info_pub.publish(right_info)

    def start(self):
        robot = sys.argv[1] + "_" + sys.argv[2]
        logging.debug("Starting Noisy Image Elimination: ")
        rospy.init_node('noisy_image_leliminate', anonymous = False)

        image_sub_right = message_filters.Subscriber('/'+robot+'/camera/right/image_raw', Image)
        image_sub_left = message_filters.Subscriber('/'+robot+'/camera/left/image_raw', Image)
        info_sub_right = message_filters.Subscriber('/'+robot+'/camera/right/camera_info', CameraInfo)
        info_sub_left = message_filters.Subscriber('/'+robot+'/camera/left/camera_info', CameraInfo)

        self.left_image_pub = rospy.Publisher('/capricorn/'+robot+'/camera/left/image_raw', Image, queue_size=10)
        self.right_image_pub = rospy.Publisher('/capricorn/'+robot+'/camera/right/image_raw', Image, queue_size=10)
        self.left_info_pub = rospy.Publisher('/capricorn/'+robot+'/camera/left/camera_info', CameraInfo, queue_size=10)
        self.right_info_pub = rospy.Publisher('/capricorn/'+robot+'/camera/right/camera_info', CameraInfo, queue_size=10)
        ts = message_filters.ApproximateTimeSynchronizer([image_sub_right, info_sub_right, image_sub_left, info_sub_left], 10, 0.5, allow_headerless=False)
        ts.registerCallback(self.loop)
        rospy.spin()

if __name__=='__main__':
    if len(rospy.myargv(argv=sys.argv)) < 3:
        print("Usage: rosrun noisy_image_eliminate.py <robot name> <robot number>")
        sys.exit()

    n = noisy_image_eliminate()
    n.start()
    

File: perception/src/noisy_image_eliminate.cpp
#include<iostream>

#include <ros/ros.h>
#include <message_filters/subscriber.h>
#include <message_filters/time_synchronizer.h>
#include <message_filters/sync_policies/approximate_time.h>
#include <sensor_msgs/Image.h>
#include <sensor_msgs/CameraInfo.h>
#include <cv_bridge/cv_bridge.h>

#include <opencv2/core/core.hpp>
#include <opencv2/opencv.hpp>

std::string robot_name;

// Initializing noisy image check parameters
float right_last_sum = 0;
float left_last_sum = 0 ;
float right_actual_sum = 0; 
float left_actual_sum = 0;
float threshold = 4000;
float upper_threshold = 1.3;
float lower_threshold = 0.7;

// Initializing global publishers
ros::Publisher right_image_pub, left_image_pub, right_info_pub, left_info_pub;

/**
 * @brief Checks whether an image is noise free or not
 * 
 * @param img ROS Image Message : Image to be check, should not be null
 * @param left Bool : Whether the image is a left image of a stereo pair
 * @return true if image does not have noise
 * @return false if image has noise
 */
bool check_image(const sensor_msgs::ImageConstPtr& img, float& actual_sum, float& last_sum) 
{
    
    cv_bridge::CvImagePtr cv_ptr;
    cv_ptr = cv_bridge::toCvCopy(img, sensor_msgs::image_encodings::BGR8);
    cv::Mat edge;
    cv::Canny(cv_ptr->image, edge, 250, 255);
    
    int s = cv::sum(edge)[0];

    int diff = s - last_sum; 
    if(s >= actual_sum * lower_threshold && s <= actual_sum * upper_threshold) 
    {
        last_sum = s;
    }

    actual_sum = s;

    if(diff < threshold) 
    {
        last_sum = s;
        return true;
    }
    else 
    {
        return false;
    }
}

/**
 * @brief : Callback function for getting messages from stereo camera, publishes images and camera info
 * on new capricorn topics if the images does not have noise
 * 
 * @param right : Right Image (ROS Image msg)
 * @param left  : Left Image (ROS Image msg)
 * @param right_info : Right Camera Info (ROS CameraInfo msg)
 * @param left_info  : Left Camera Info (ROS CameraInfo msg)
 */
void img_callback(const sensor_msgs::ImageConstPtr& right, const sensor_msgs::ImageConstPtr& left, const sensor_msgs::CameraInfoConstPtr& right_info, const sensor_msgs::CameraInfoConstPtr& left_info) 
{
    if(check_image(right, right_actual_sum, right_last_sum)) 
    {
        right_image_pub.publish(right);
        right_info_pub.publish(right_info);
    }
    
    if(check_image(left, left_actual_sum, left_last_sum)) 
    {
        left_image_pub.publish(left);
        left_info_pub.publish(left_info);
    }
}

int main(int argc, char *argv[]) 
{
    std::string name(argv[1]);
    robot_name = name;

    ros::init(argc, argv, "noisy_image_eliminate_"+robot_name);  std::cout<<"9 7  \n";
    ros::NodeHandle nh("");

    message_filters::Subscriber<sensor_msgs::Image> img_sub_r(nh, '/'+robot_name+"/camera/right/image_raw", 1);
    message_filters::Subscriber<sensor_msgs::Image> img_sub_l(nh, '/'+robot_name+"/camera/left/image_raw", 1);
    message_filters::Subscriber<sensor_msgs::CameraInfo> info_sub_r(nh, '/'+robot_name+"/camera/right/camera_info", 1);
    message_filters::Subscriber<sensor_msgs::CameraInfo> info_sub_l(nh, '/'+robot_name+"/camera/left/camera_info", 1);
    
    typedef message_filters::sync_policies::ApproximateTime<sensor_msgs::Image, sensor_msgs::Image, sensor_msgs::CameraInfo, sensor_msgs::CameraInfo> syncPolicy;
    message_filters::Synchronizer<syncPolicy> sync(syncPolicy(10), img_sub_r, img_sub_l, info_sub_r, info_sub_l);
    sync.registerCallback(boost::bind(&img_callback, _1, _2, _3, _4));
    
    right_image_pub = nh.advertise<sensor_msgs::Image>("/capricorn/" + robot_name + "/camera/right/image_raw", 10);
    left_image_pub = nh.advertise<sensor_msgs::Image>("/capricorn/" + robot_name + "/camera/left/image_raw", 10);
    right_info_pub = nh.advertise<sensor_msgs::CameraInfo>("/capricorn/" + robot_name + "/camera/right/camera_info", 10);
    left_info_pub = nh.advertise<sensor_msgs::CameraInfo>("/capricorn/" + robot_name + "/camera/left/camera_info", 10);

    ros::spin();
}